src/pivot-table/hooks/use-items-rendered-handler/index.ts

```diff
@@ -0,0 +1,74 @@
+/*  eslint-disable no-param-reassign */
+import { useCallback } from "react";
+import { type GridOnItemsRenderedProps } from "react-window";
+import type { DataModel, LayoutService, MeasureData, PageInfo, ViewService } from "../../../types/types";
+import useScrollDirection from "../use-scroll-direction";
+import { BUFFER, debouncedFetchPages, throttledFetchPages } from "./utils/fetch-pages";
+
+type Props = {
+  viewService: ViewService;
+  layoutService: LayoutService;
+  dataModel: DataModel;
+  measureData: MeasureData;
+  pageInfo: PageInfo;
+};
+
+const DEBOUNCED_GRID_SIZE_THRESHOLD = 1500;
+
+const useItemsRenderedHandler = ({ viewService, dataModel, measureData, layoutService, pageInfo }: Props) => {
+  const { scrollHandler, verticalScrollDirection, horizontalScrollDirection } = useScrollDirection();
+
+  /**
+   * react-window callback that is called when the range of items rendered by the VariableSizeGrid changes.
+   *
+   * It's intended to handle the following scenarions that might require additional data to be fetch:
+   * - Scrolling
+   * - Re-sizing the chart
+   * - Theme change (ex: go from large font-size to small could change the number of rendered cells)
+   */
+  const onItemsRenderedHandler = useCallback(
+    async ({
+      overscanColumnStartIndex,
+      overscanColumnStopIndex,
+      overscanRowStartIndex,
+      overscanRowStopIndex,
+      visibleColumnStartIndex,
+    }: GridOnItemsRenderedProps) => {
+      viewService.gridColumnStartIndex = visibleColumnStartIndex;
+      viewService.gridRowStartIndex = overscanRowStartIndex;
+      viewService.gridWidth = overscanColumnStopIndex - overscanColumnStartIndex + 1;
+      viewService.gridHeight = overscanRowStopIndex - overscanRowStartIndex + 1;
+
+      let throttledOrDebouncedFetchPages = throttledFetchPages;
+
+      /**
+       * A throttled fetch gives the best user experience as it reduces the number of empty cells the
+       * user sees. The debounced fetch greatly improves performance at the cost of the user experience.
+       *
+       * This is a control mechanism that allows us to tweak where the user experience vs performance
+       * threshold should be.
+       */
+      if (viewService.gridWidth * (viewService.gridHeight + BUFFER) > DEBOUNCED_GRID_SIZE_THRESHOLD) {
```

Comment:
When fetching data the `BUFFER` is only applied in the scroll direction. Which is why I also only added it here in one direction. As the idea was that `viewService.gridWidth * (viewService.gridHeight + BUFFER)` should be roughly equal to the amount of that that will be fetched.

But looking at the logic again. I think i'st faulty as `10 * (5 + BUFFER) !== (10 + BUFFER) * 5`. So I should at least add buffer based on the scroll direction.

Reply:
I think you are correct regarding only using cells from the data grid. That will not always give a fair value. Will look into that. Good find!

Example:
Left grid has 1000 dimension that are all expanded. Data grid have 1 column and 50 rows. With current logic that would be 1 * 50. But in reality we would fetch (1000+1) * 50 cells.

Reply:
> When fetching data the BUFFER is only applied in the scroll direction.

currently only for Y direction right? (the grid height + buffer part)

> So I should at least add buffer based on the scroll direction.

Maybe I'm a bit crazy on this + it might sound a bit hard to implement, but what if we keep predicting user behaviour in this case? 
like if user makes two (or n) scroll(s) in the same direction, then we can say that the next scroll direction is going to be the same as previous ones (maybe because user tries scroll continuously in that direction) -> so next time while fetching data -> apply more buffer in that direction.

OR since we are in the right track on guessing user's scroll direction -> may be it would be nice to apply buffer with a **bigger coefficient** than previous times to get more data upfront, like: 
first time guessing correct -> apply `buffer` amount, 
second time apply `buffer * 1.15` amount, 
third time apply `buffer * 1.3`  and ...
**until user stops scrolling** in same direction and our prediction process stops.

Not sure if that is a good idea or not 🤷‍♂️

Reply:
Not a bad idea per say, but the generally the trade-off is user experiance (more buffer) vs performance (less buffer/less data fetched). So we can't just keep adding to the buffer without a cost in performance.